packages/ui-scripts/src/publish.ts

```diff
@@ -62,73 +65,140 @@ async function publish({
+    return publishSnapshotVersion({
+      version,
+      packageName,
+      packages
+    })
+  }
+}
+
+/**
+ * Publishes each package to npm.
+ */
+async function publishRegularVersion(arg: {
+  version: string
+  tag: string
+  packages: PackageInfo[]
+}) {
+  const { version, packages, tag } = arg
+  for await (const pkg of publishPackages(packages, version, tag)) {
+    info(`📦  Version ${version} of ${pkg.name} was successfully published!`)
+  }
+}
+
+/**
+ * Calculates the new snapshot version based on the latest tag
+ * and the current commit, then publishes each package to npm
+ * with the new snapshot version.
+ */
+async function publishSnapshotVersion(arg: {
+  version: string
+  packageName: string
+  packages: PackageInfo[]
+}) {
+  const { version, packageName, packages } = arg
+  const snapshotVersion = calculateNextSnapshotVersion(version)
+
+  info(`applying new snapshot version (${snapshotVersion}) to each package`)
+
+  await bumpPackages(packageName, snapshotVersion)
+
+  for await (const pkg of publishPackages(packages, snapshotVersion)) {
+    info(
+      `📦  Version ${snapshotVersion} of ${pkg.name} was successfully published!`
+    )
+  }
+}
+
+/**
+ * Calculates the new snapshot version.
+ * @returns the new snapshot version
+ */
+function calculateNextSnapshotVersion(version: string): string {
+  const ver = `v${version}`
+  // get the commit count between the current released version
+  // and the commit that we are on currently
+  // note: this will not include the release commit into the count
+  const commitCount = runGitCommand([
+    'rev-list',
+    '--count',
+    `${ver}..${runGitCommand(['describe'])}`
+  ])
+
+  // we have to decrease the commitCount by 1 because the snapshots are
+  // zero based
+  const snapshotVersion = `${semver.inc(version, 'patch')}-snapshot-${
+    Number(commitCount) - 1
+  }`
+
+  info(`next snapshot version is: ${snapshotVersion}`)
+
+  return snapshotVersion
+}
+
+/**
+ * An async generator function which will do the publishing
+ * for each pacakge.
+ */
+async function* publishPackages(
+  packages: PackageInfo[],
+  version: string,
+  tag?: string
+) {
+  const wait = (delay: number) =>
+    new Promise((resolve) => {
+      setTimeout(resolve, delay)
+    })
+
+  for (const pkg of packages) {
+    const { stdout } = await runCommandAsync(
+      'npm',
+      ['info', pkg.name, '--json'],
+      [],
+      {
+        stdio: 'pipe'
+      }
+    )
+    const packageVersions: string[] = JSON.parse(stdout).versions
+
+    // if the package is already in the registry then don't try
+    // to publish it again
+    if (packageVersions.includes(version)) {
+      info(`📦  v${version} of ${pkg.name} is already published!`)
+    } else {
+      const publishArgs = ['publish', pkg.location]
+      if (tag) {
+        publishArgs.push(...['--tag', tag])
+      }
+      await runCommandAsync('npm', publishArgs, [])
+
+      await wait(500)
+
+      yield pkg
     }
   }
 }
```

Comment:
This is the main chunk of the change, this async generator function will release each package that is given to it with the correct version and will wait 500ms between releasing packages.